core/controllers/resources_test.py

```diff
@@ -133,3 +132,45 @@ def test_bad_filenames_are_detected(self):
         self.assertIn('Filenames should not include', response_dict['error'])
 
         self.logout()
+
+    def test_missing_extensions_are_detected(self):
```

Comment:
After talking about this offline, I've added one more test, which passes PNG data but with a JPG filename. The test verifies that the image is successfully uploaded and stored as JPG.

Even though this seems counterintuitive as a test (we're letting an incorrect case 'succeed') it ensures that the system is not using binary data test to figure out extension, but the filename instead. This was a design decision based on the false negative found by the imghdr.what() method.